### PR TITLE
feat: consume all fees

### PR DESCRIPTION
**Pull request by nowooj**
- consume all fees

### CODE REVIEW

**Comment by jhyeom26** [comment]
Would be good if we leave a comment that we also transfer the change after execution to a validator on purpose as our fair system.

**Review by yoosah — APPROVED** [review]

**Review by psy2848048 — APPROVED** [review]

File: execution-engine/engine-tests/src/test/hdac_system_contracts/client_api_proxy.rs [hunk 1]
```diff
@@ -200,20 +196,7 @@ fn should_invoke_successful_bond_and_unbond() {
...
-    let pos_transform = &transforms[&Key::from(pos_uref).normalize()];
-
-    let pos_contract = if let Transform::Write(StoredValue::Contract(contract)) = pos_transform {
-        contract
-    } else {
-        panic!(
-            "pos transform is expected to be of AddKeys variant but received {:?}",
-            pos_transform
-        );
-    };
```
Comment (jhyeom26) [inline execution-engine/engine-tests/src/test/hdac_system_contracts/client_api_proxy.rs:216]: Why this code is deleted~? The replaced code has a slightly different meaning.
This asserts whether the bonding_request includes Transform::Write.

File: execution-engine/engine-tests/src/test/hdac_system_contracts/client_api_proxy.rs [hunk 2]
```diff
@@ -129,16 +129,12 @@ fn should_invoke_successful_standard_payment() {
         "balance should be less than initial balance"
     );
 
-    let response = transfer_result
+    transfer_result
         .builder()
         .get_exec_response(0)
-        .expect("there should be a response")
-        .clone();
+        .expect("there should be a response");
```
Comment (jhyeom26) [inline execution-engine/engine-tests/src/test/hdac_system_contracts/client_api_proxy.rs]: If your intention is to discard the returned value, `let _ = transfer_result` would be more explicit.

File: execution-engine/engine-tests/src/test/hdac_system_contracts/client_api_proxy.rs [hunk 3]
```diff
@@ -235,17 +218,7 @@ fn should_invoke_successful_bond_and_unbond() {
...
-    let pos_transform = &transforms[&Key::from(pos_uref).normalize()];
-
-    let pos_contract = if let Transform::Write(StoredValue::Contract(contract)) = pos_transform {
-        contract
-    } else {
-        panic!(
-            "pos transform is expected to be of AddKeys variant but received {:?}",
-            pos_transform
-        );
-    };
```
Comment (jhyeom26) [inline execution-engine/engine-tests/src/test/hdac_system_contracts/client_api_proxy.rs:248]: Same here.

File: execution-engine/engine-tests/src/test/hdac_system_contracts/proof_of_stake/bonding.rs [hunk 4]
```diff
@@ -199,15 +199,13 @@ fn should_run_successful_bond_and_unbond() {
         .finish();
 
     let account_1_bal_after = builder.get_purse_balance(account_1.main_purse());
-    let exec_response = builder
+    builder
```
Comment (jhyeom26) [inline execution-engine/engine-tests/src/test/hdac_system_contracts/proof_of_stake/bonding.rs]: `let _ =` same here

File: execution-engine/engine-tests/src/test/hdac_system_contracts/proof_of_stake/bonding.rs [hunk 5]
```diff
@@ -299,15 +295,13 @@ fn should_run_successful_bond_and_unbond() {
...
```
Comment (jhyeom26) [inline execution-engine/engine-tests/src/test/hdac_system_contracts/proof_of_stake/bonding.rs]: same here

File: execution-engine/engine-tests/src/test/hdac_system_contracts/proof_of_stake/bonding.rs [hunk 6]
```diff
@@ -346,21 +340,16 @@ fn should_run_successful_bond_and_unbond() {
         .commit()
         .finish();
 
-    let exec_response = builder
+    builder
```
Comment (jhyeom26) [inline execution-engine/engine-tests/src/test/hdac_system_contracts/proof_of_stake/bonding.rs]: same here

File: execution-engine/engine-tests/src/test/hdac_system_contracts/proof_of_stake/delegation.rs [hunk 7]
```diff
@@ -306,21 +298,16 @@ fn should_run_successful_delegate_and_undelegate() {
     );
 
     // validate ACCOUNT_2's balance
-    let undelegate_all_response = builder
+    builder
```
Comment (jhyeom26) [inline execution-engine/engine-tests/src/test/hdac_system_contracts/proof_of_stake/delegation.rs]: same here

File: execution-engine/engine-tests/src/test/hdac_system_contracts/proof_of_stake/delegation.rs [hunk 8]
```diff
@@ -537,21 +524,18 @@ fn should_run_successful_redelegate() {
     );
 
     // validate ACCOUNT_3's balance
-    let undelegate_all_response = builder
+    builder
```
Comment (jhyeom26) [inline execution-engine/engine-tests/src/test/hdac_system_contracts/proof_of_stake/delegation.rs]: same here

File: execution-engine/engine-tests/src/test/hdac_system_contracts/proof_of_stake/finalize_payment.rs [hunk 9]
```diff
@@ -121,14 +118,11 @@ fn finalize_payment_should_refund_to_specified_purse() {
     builder.exec(exec_request).expect_success().commit();
 
     let spent_amount: U512 = {
-        let response = builder
+        builder
```
Comment (jhyeom26) [inline execution-engine/engine-tests/src/test/hdac_system_contracts/proof_of_stake/finalize_payment.rs]: same here

File: execution-engine/engine-tests/src/test/hdac_system_contracts/standard_payment.rs [hunk 10]
```diff
@@ -418,11 +418,7 @@ fn should_correctly_charge_when_session_code_runs_out_of_gas() {
         .get_exec_response(0)
         .expect("there should be a response");
 
-    let success_result = utils::get_success_result(&response);
-    let gas = success_result.cost();
-    let motes = Motes::from_gas(gas, CONV_RATE).expect("should have motes");
-
-    let tally = motes.value() + modified_balance;
+    let tally = modified_balance + *DEFAULT_PAYMENT;
```
Comment (jhyeom26) [inline execution-engine/engine-tests/src/test/hdac_system_contracts/standard_payment.rs]: `let tally = modified_balance + payment_purse_amount;` seems more semantic.

File: execution-engine/engine-tests/src/test/hdac_system_contracts/standard_payment.rs [hunk 11]
```diff
@@ -476,15 +472,11 @@ fn should_correctly_charge_when_session_code_fails() {
         "balance should be less than initial balance"
     );
 
-    let response = builder
+    builder
```
Comment (jhyeom26) [inline execution-engine/engine-tests/src/test/hdac_system_contracts/standard_payment.rs]: `let _ =` same here

File: execution-engine/engine-tests/src/test/hdac_system_contracts/standard_payment.rs [hunk 12]
```diff
@@ -476,15 +472,11 @@ fn should_correctly_charge_when_session_code_fails() {
...
         .get_exec_response(0)
-        .expect("there should be a response")
-        .clone();
+        .expect("there should be a response");
 
-    let success_result = utils::get_success_result(&response);
-    let gas = success_result.cost();
-    let motes = Motes::from_gas(gas, CONV_RATE).expect("should have motes");
-    let tally = motes.value() + modified_balance;
+    let tally = modified_balance + *DEFAULT_PAYMENT;
```
Comment (jhyeom26) [inline execution-engine/engine-tests/src/test/hdac_system_contracts/standard_payment.rs]: `let tally = modified_balance + payment_purse_amount;` seems more semantic.

File: execution-engine/engine-tests/src/test/hdac_system_contracts/standard_payment.rs [hunk 13]
```diff
@@ -534,15 +526,11 @@ fn should_correctly_charge_when_session_code_succeeds() {
         "balance should be less than initial balance"
     );
 
-    let response = builder
+    builder
```
Comment (jhyeom26) [inline execution-engine/engine-tests/src/test/hdac_system_contracts/standard_payment.rs]: `let _ =` same here

File: execution-engine/engine-tests/src/test/hdac_system_contracts/standard_payment.rs [hunk 14]
```diff
@@ -534,15 +526,11 @@ fn should_correctly_charge_when_session_code_succeeds() {
...
         .get_exec_response(0)
-        .expect("there should be a response")
-        .clone();
+        .expect("there should be a response");
 
-    let success_result = utils::get_success_result(&response);
-    let gas = success_result.cost();
-    let motes = Motes::from_gas(gas, CONV_RATE).expect("should have motes");
-    let total = motes.value() + U512::from(transferred_amount);
+    let total = U512::from(transferred_amount) + *DEFAULT_PAYMENT;
```
Comment (jhyeom26) [inline execution-engine/engine-tests/src/test/hdac_system_contracts/standard_payment.rs]: same here

File: execution-engine/engine-tests/src/test/hdac_system_contracts/client_api_proxy.rs (hunk 2 above)
Comment (nowooj) [reply execution-engine/engine-tests/src/test/hdac_system_contracts/client_api_proxy.rs]: https://github.com/hdac-io/CasperLabs/pull/23/commits/2df36d4354b27b93666538c9bd57b1d8e1f05edb

File: execution-engine/engine-tests/src/test/hdac_system_contracts/proof_of_stake/bonding.rs (hunk 4 above)
Comment (nowooj) [reply execution-engine/engine-tests/src/test/hdac_system_contracts/proof_of_stake/bonding.rs]: https://github.com/hdac-io/CasperLabs/pull/23/commits/2df36d4354b27b93666538c9bd57b1d8e1f05edb

File: execution-engine/engine-tests/src/test/hdac_system_contracts/proof_of_stake/bonding.rs (hunk 5 above)
Comment (nowooj) [reply execution-engine/engine-tests/src/test/hdac_system_contracts/proof_of_stake/bonding.rs]: https://github.com/hdac-io/CasperLabs/pull/23/commits/2df36d4354b27b93666538c9bd57b1d8e1f05edb

File: execution-engine/engine-tests/src/test/hdac_system_contracts/standard_payment.rs (hunk 10 above)
Comment (nowooj) [reply execution-engine/engine-tests/src/test/hdac_system_contracts/standard_payment.rs]: https://github.com/hdac-io/CasperLabs/pull/23/commits/2df36d4354b27b93666538c9bd57b1d8e1f05edb

File: execution-engine/engine-tests/src/test/hdac_system_contracts/proof_of_stake/bonding.rs (hunk 6 above)
Comment (nowooj) [reply execution-engine/engine-tests/src/test/hdac_system_contracts/proof_of_stake/bonding.rs]: https://github.com/hdac-io/CasperLabs/pull/23/commits/2df36d4354b27b93666538c9bd57b1d8e1f05edb

File: execution-engine/engine-tests/src/test/hdac_system_contracts/proof_of_stake/delegation.rs (hunk 7 above)
Comment (nowooj) [reply execution-engine/engine-tests/src/test/hdac_system_contracts/proof_of_stake/delegation.rs]: https://github.com/hdac-io/CasperLabs/pull/23/commits/2df36d4354b27b93666538c9bd57b1d8e1f05edb

File: execution-engine/engine-tests/src/test/hdac_system_contracts/proof_of_stake/delegation.rs (hunk 8 above)
Comment (nowooj) [reply execution-engine/engine-tests/src/test/hdac_system_contracts/proof_of_stake/delegation.rs]: https://github.com/hdac-io/CasperLabs/pull/23/commits/2df36d4354b27b93666538c9bd57b1d8e1f05edb

File: execution-engine/engine-tests/src/test/hdac_system_contracts/proof_of_stake/finalize_payment.rs (hunk 9 above)
Comment (nowooj) [reply execution-engine/engine-tests/src/test/hdac_system_contracts/proof_of_stake/finalize_payment.rs]: https://github.com/hdac-io/CasperLabs/pull/23/commits/2df36d4354b27b93666538c9bd57b1d8e1f05edb

File: execution-engine/engine-tests/src/test/hdac_system_contracts/standard_payment.rs (hunk 11 above)
Comment (nowooj) [reply execution-engine/engine-tests/src/test/hdac_system_contracts/standard_payment.rs]: https://github.com/hdac-io/CasperLabs/pull/23/commits/2df36d4354b27b93666538c9bd57b1d8e1f05edb

File: execution-engine/engine-tests/src/test/hdac_system_contracts/standard_payment.rs (hunk 12 above)
Comment (nowooj) [reply execution-engine/engine-tests/src/test/hdac_system_contracts/standard_payment.rs]: https://github.com/hdac-io/CasperLabs/pull/23/commits/2df36d4354b27b93666538c9bd57b1d8e1f05edb

File: execution-engine/engine-tests/src/test/hdac_system_contracts/standard_payment.rs (hunk 13 above)
Comment (nowooj) [reply execution-engine/engine-tests/src/test/hdac_system_contracts/standard_payment.rs]: https://github.com/hdac-io/CasperLabs/pull/23/commits/2df36d4354b27b93666538c9bd57b1d8e1f05edb

File: execution-engine/engine-tests/src/test/hdac_system_contracts/standard_payment.rs (hunk 14 above)
Comment (nowooj) [reply execution-engine/engine-tests/src/test/hdac_system_contracts/standard_payment.rs]: https://github.com/hdac-io/CasperLabs/pull/23/commits/2df36d4354b27b93666538c9bd57b1d8e1f05edb

File: execution-engine/engine-tests/src/test/hdac_system_contracts/client_api_proxy.rs (hunk 1 above)
Comment (jhyeom26) [reply execution-engine/engine-tests/src/test/hdac_system_contracts/client_api_proxy.rs:216]: As we discussed in slack, resolve this conversation.

**Comment by jhyeom26** [comment]
How about deleting all unused response which was required for the gas consumption calculation?
These response assertions are actually checked implicitly or meaningless.

For example,
```rust
let spent_amount: U512 = {
    let _ = builder
        .get_exec_response(0)
        .expect("there should be a response");
   *DEFAULT_PAYMENT
};
```

**Comment by nowooj** [comment]
> How about deleting all unused response which was required for the gas consumption calculation?
> These response assertions are actually checked implicitly or meaningless.
> 
> For example,
> 
> ```rust
> let spent_amount: U512 = {
>     let _ = builder
>         .get_exec_response(0)
>         .expect("there should be a response");
>    *DEFAULT_PAYMENT
> };
> ```

@jhyeom26 
It looks like nothing except finalize_payment. Modify that test. https://github.com/hdac-io/CasperLabs/pull/23/commits/859759895324d22a730607b8f150e2abf4825567

**Comment by jhyeom26** [comment]
I meant all of them

```rust
let _ = builder
    .get_exec_response(0)	
    .expect("there should be a response");
```

Checking only the existence of `exec_response` is already included in `expect_success()`.
Even in the tests not calling expect_success, they inspect the value of the failure which means unwrapping the response.

**Comment by nowooj** [comment]
> I meant all of them
> 
> ```rust
> let _ = builder
>     .get_exec_response(0)	
>     .expect("there should be a response");
> ```
> 
> Checking only the existence of `exec_response` is already included in `expect_success()`.
> Even in the tests not calling expect_success, they inspect the value of the failure which means unwrapping the response.

Ah... Complete to here. https://github.com/hdac-io/CasperLabs/commit/a6c787b3ab3ee54efa91df895cfbe1dba927397c

**Review by jhyeom26 — APPROVED** [review]

**Pull request merged**